Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -135,18 +134,29 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
                 self.type = .ipLiteral
             } else {
                 let parts = hostAndPort.components(separatedBy: ":")
-                self.host = String(parts[0])
                 guard parts.count <= 2 else {
                     throw MongoError.InvalidArgumentError(
                         message: "expected only a single port delimiter ':' in \(hostAndPort)"
                     )
                 }
+
+                let host = parts[0]
+                if host.hasSuffix(".sock") {
+                    self.host = try host.getPercentDecoded(forKey: "UNIX domain socket")
+                    self.type = .unixDomainSocket
+                } else if host.isIPv4() {
```

Comment:
The logic here is based off of this language in [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2):

> `host        = IP-literal / IPv4address / reg-name`
>
>The syntax rule for host is ambiguous because it does not completely
distinguish between an IPv4address and a reg-name.  In order to
disambiguate the syntax, we apply the "first-match-wins" algorithm:
If host matches the rule for IPv4address, then it should be
considered an IPv4 address literal and not a reg-name.

Essentially, if the host does not match IPv4, fall back to hostname. The rules for reg-name permit "invalid" IPv4 addresses, so we don't want to error on a string like "1.2.3" as that is a valid hostname.

Reply:
thanks for the explanation! and I was going to suggest we add a test to make sure we don't break this in the future but I see you already did 🙂